cfg.cpp

```diff
@@ -18,22 +18,91 @@ static droption_t<bool> no_cti
 (DROPTION_SCOPE_CLIENT, "no_cti", false,
  "Don't count control transfer instructions", "");
 
-static droption_t<std::string> output
+droption_t<std::string> output
 (DROPTION_SCOPE_CLIENT, "output", "",
  "Output results to file", "");
 
+droption_t<bool> txt
+(DROPTION_SCOPE_CLIENT, "txt", false,
+ "Print text output while running instead of json at the end of the run - you will have to take care of the doubles youself", "");
+
+static void *mutex;
+static int tcls_idx;
+static FILE *out = stdout;
+
+typedef struct {
+  void *addr;
+} per_thread_t;
+
+
 void
 dr_exit(void)
 {
-    if (output.get_value() == "")
-        std::cout << std::setw(2) << construct_json() << std::endl;
-    else {
-        std::ofstream ofs(output.get_value());
-        ofs << std::setw(2) << construct_json() << std::endl;
+    if (!txt.get_value()) {
+        if (output.get_value() == "")
+            std::cout << std::hex << std::setw(2) << construct_json() << std::endl;
+        else {
+            std::ofstream ofs(output.get_value());
+            ofs << std::hex << std::setw(2) << construct_json() << std::endl;
+        }
     }
     drmgr_exit();
 }
 
+void 
+event_thread_context_init(void *drcontext, bool new_depth)
+{
+    per_thread_t *data;
+
+    if (new_depth) {
+        data = (per_thread_t *) dr_thread_alloc(drcontext, sizeof(per_thread_t));
+        drmgr_set_cls_field(drcontext, tcls_idx, data);
+    } else {
+        data = (per_thread_t *) drmgr_get_cls_field(drcontext, tcls_idx);
+    }
+    memset(data, 0, sizeof(*data));
+}
+
+void
+event_thread_context_exit(void *drcontext, bool thread_exit)
+{
+    if (thread_exit) {
+        per_thread_t *data = (per_thread_t *) drmgr_get_cls_field(drcontext, tcls_idx);
+        dr_thread_free(drcontext, data, sizeof(per_thread_t));
+        drmgr_set_cls_field(drcontext, tcls_idx, NULL);
+    }
+}
+
+dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                      bool for_trace, bool translating, void *user_data)
+{
+    per_thread_t *data = (per_thread_t *) drmgr_get_cls_field(drcontext, tcls_idx);
```

Comment:
It looks like this is trying to re-implement cbr.cpp. I would not recommend this, as it'll probably be easier to fix cbr.cpp and adapt it for ARM (and potentially re-contribute the fix towards the cbr.c sample in the DynamoRIO repo which should exhibit the same problems). 

If you do not have the time for this, I can take a stab at it within the next week or so.

Reply:
Another thing--I'm wondering if this works in general. DynamoRIO fires an `event_app_instruction` event the first time it sees an instruction, but not subsequently. Consider the following control flow:

```
  a
 / \
b   c
 \ /
  d
```

It's a simple if-then-else statement in C. However, if we record the edge `b-d`, then on a subsequent execution from path `a-c-d` I wouldn't expect the DynamoRIO event to fire when we execute the instructions in basic block `d`. Does this work for you?

Reply:
just reading this after reading & answering your email - just returned from vacation :)

actually, you have a point that your a -> b|c ->d might only log one path. that might be the case.
I will investigate and give you a response on that. if it is like you think, then yes, my approach is broken. if it work like I hope or thought it would work, then it would be the easiest solution.

Reply:
OK, done.

I did a comparison to drcfg in your repository (called "before.txt") and my cloned repository with the changes (called "after.txt").  I disabled random address space, and traced "unrar -inul p test.rar"

```
# ls -l after.txt before.txt
-rw-r--r-- 1 root root 324991 Apr 16 16:58 after.txt
-rw-r--r-- 1 root root 325064 Apr 16 16:58 before.txt
# wc -l after.txt before.txt
 20320 after.txt
 20325 before.txt
 40645 total
```

if your assumption would hold true, then for every source, there would only be one target recorded. however I find in the my patched results even results with multiple entries, e.g.:
```
      "address": 140737351486947,
      "targets": [
        140737324874432,
        4205056,
        140737333003728,
        140737329276560,
        140737327079680
      ]
```

so it looks like my approach is working. I would feel more sure if you would verify with your own tests though :)

Reply:
Sure, will take a look w/in the next few days :)

Reply:
Consider the following code: [cbr.zip](https://github.com/toshipiazza/drcfg/files/1918645/cbr.zip)

The CFG looks something like: 

![capture](https://user-images.githubusercontent.com/6254422/38856403-650e4f0e-41f4-11e8-80c2-e7046c2d6e36.PNG)

It has the diamond shape, like I mentioned above. This PR yields the following output:

```
{
  "branches": [
    {
      "address": 40008a,
      "targets": [
        40008c
      ]
    },
    {
      "address": 400081,
      "targets": [
        400085,
        400083
      ]
    }
  ]
}
```

Clearly we're missing one of the targets (to the done branch, I think).